js/stdlib/ts/ALGO.ts

```diff
@@ -589,20 +589,83 @@ type QueryResult =
   | { succ: true, txn: any }
   | { succ: false, round: number }
 
-const doQuery = async (dhead:string, query: ApiCall<any>, pred: ((x:any) => boolean) = ((x) => { void(x); return true; })): Promise<QueryResult> => {
-  const res = await doQuery_(dhead, query);
-  const txns = res.transactions;
-  const ptxns = txns.filter(pred);
-  debug(dhead, {ptxns});
-
-  if ( ptxns.length == 0 ) {
-    return { succ: false, round: res['current-round'] };
+// ****************************************************************************
+// Event Cache
+// ****************************************************************************
+
+export const chooseMinRoundTxn = (ptxns: any[]) =>
+  ptxns.reduce((accum: any, x: any) =>
+      (x['confirmed-round'] < accum['confirmed-round']) ? x : accum, ptxns[0])
+
+export type RoundInfo = {
+  minRound?: number,
+  maxRound?: number,
+  specRound?: number,
+}
+
+export class EventCache {
+
+  cache: any[] = [];
+
+  currentRound = 0;
+
+  constructor() {
+    this.cache = [];
   }
 
-  const txn = ptxns.reduce((accum: any, x: any) =>
-    (x['confirmed-round'] < accum['confirmed-round']) ? x : accum, ptxns[0]);
-  return { succ: true, txn };
-};
+  async query(dhead: string, ApplicationID: number, roundInfo: RoundInfo, pred: ((x:any) => boolean)): Promise<QueryResult> {
+    const { minRound, maxRound, specRound } = roundInfo;
+    debug(dhead, `EventCache.query`, ApplicationID, minRound, specRound, maxRound, this.currentRound);
+
+    // Clear cache of stale transactions
+    const filterRound = minRound || specRound || 0;
+    this.cache = this.cache.filter(x => x['confirmed-round'] > filterRound);
+
+    // Check to see if the transaction we want is in cache
+    const initPtxns = this.cache.filter(pred);
+    if (initPtxns.length != 0) {
+      debug(`Found transaction in Event Cache`);
+      const txn = chooseMinRoundTxn(initPtxns)
+      this.currentRound = txn['confirmed-round'];
+      return { succ: true, txn };
+    }
+
+    debug(`Transaction not in Event Cache. Querying network...`);
```

Comment:
This is not right either, because it's not that we ask the network when the query isn't answered, it is if the range of the query is not included in the cache. For example, if the cache holds `[100, 200]`, but we are looking for something in the range `[150, 1000]`, then we need to got to the network to get more txns. Now, we could first check `[150, 200]` in the cache, then if that fails, go to the network for `[200, 1000]`

Reply:
Fixed this to work as you described